invenio_users_resources/records/hooks.py

```diff
@@ -52,15 +50,18 @@ def post_commit(sender, session):
     # DB operations are allowed here, not even lazy-loading of
     # properties!
     sid = id(session)
-    for user_id in current_db_change_history.updated_users[sid]:
-        reindex_user.delay(user_id)
+    if current_db_change_history.sessions.get(sid):
+        for user_id in current_db_change_history.sessions[sid].updated_users:
+            reindex_user.delay(user_id)
 
-    for role_id in current_db_change_history.updated_roles[sid]:
-        reindex_group.delay(role_id)
+        for role_id in current_db_change_history.sessions[sid].updated_roles:
+            reindex_group.delay(role_id)
 
-    for user_id in current_db_change_history.deleted_users[sid]:
-        unindex_user.delay(user_id)
-    for role_id in current_db_change_history.deleted_roles[sid]:
-        unindex_group.delay(role_id)
+        for user_id in current_db_change_history.sessions[sid].deleted_users:
+            unindex_user.delay(user_id)
 
-    current_db_change_history._clear_dirty_sets(session)
+        for role_id in current_db_change_history.sessions[sid].deleted_roles:
+            unindex_group.delay(role_id)
+
+        current_db_change_history.sessions[sid].indexed = True
```

Comment:
this is not guaranteed, given that it will happen in celery tasks? It also feels like that this is an extra state, added because otherwise the session is cleared in another wrong cases...

Reply:
No, these are custom signal handlers, not executed by celery tasks. From the [docs](https://docs.python.org/3.4/library/signal.html#execution-of-python-signal-handlers) looks like it should be executed always in the main thread, therefore it's safe/guaranteed. 
Yes, we have added that flag exactly because of the session risking being cleared wrongly. These are signals and we need to register 2 custom signal handlers, one is for indexing the records (this one) and another one for clearing the identity cache (in [invenio-communities](https://github.com/inveniosoftware/invenio-communities/pull/945)). However we **only** want to clear the session object when both signals were executed. Since they are signals I don't see a better way to do it, any suggstion is welcomed!